### ISSUE TITLE
Auto equation numbering

### ISSUE THREAD

**Issue by wujinq** [yzhang-gh/vscode-markdown#904]
## Proposal

Automatical equation numbering is a handy feature in technical writing, but seems unimplemented in the current version. This could be implemented in KaTeX and indeed has been proposed for several times (see [here](https://github.com/KaTeX/KaTeX/issues/350) and [here](https://github.com/KaTeX/KaTeX/issues/2003) ) but no further progress is made. 

Therefore it may be a good idea to add some sort of equation numbering and reference mechanism to Markdown All in One. My proposal is to use `[equation id]` to refer to an equation, and make equations like
```
$$
$$[equation id]
```
automatically numbered (with a tag like (1) on the rightmost place when displayed). There can be, of course, better syntax for the purpose.

**Comment by yzhang-gh** [comment]
It looks more like an extension to the Markdown syntax (similar to [Markdown footnotes](https://github.com/markdown-it/markdown-it-footnote)). I think KaTeX is still the best place to do it, which will be used by `markdown-it-katex` and then used by this extension.

**Comment by wujinq** [comment]
Yes, and this is exactly the solution used in MathJax-based Markdown math environments, in which `\eqref` and `\label` are used just like the standard LaTeX. Unfortunately, the KaTeX team seems reluctant to introduce things outside the _math_ part of LaTeX, but an equation numbering mechanism unavoidably involves with some global counting mechanisms, which cannot be tackled if KaTeX only deals with separated math blocks.

So big updates to existing systems must appear here or there: either KaTeX introduces a global counting mechanism (which deviates from their current principles), or an extension to the Markdown syntax is adopted. Or we may just have to give up equation numbering ...

**Comment by Lemmingh** [comment]
If you need complex LaTeX features, I'd recommend that you **move away from Markdown and directly write in LaTeX**.

---

There are already **too many variants/dialects/flavors** about embedding math into Markdown, causing determining semantics extremely (and probably unnecessarily) complex. Please **do not try to add another one to the chaos**.

Many users ask for lifting style restriction, and allowing custom delimiter and custom math compiler. So we **technically** cannot create another dialect.

This product is working hard to be a [general-purpose solution](https://markdown-all-in-one.github.io/docs/decisions/markdown-syntax-and-flavors.html). We do not want to tie users to a specific flavor.

---

I guess we can try to auto-complete more LaTeX commands someday. But please **do not hold much hope**. I can see a big **challenge** in synchronizing language server states.

**Comment by yzhang-gh** [comment]
> an equation numbering mechanism unavoidably involves with some global counting mechanisms

In VS Code, Markdown is rendered by the [`markdown-it`](https://github.com/markdown-it/markdown-it) engine. This repo (Markdown All in One) doesn't introduce any new Markdown syntax but only uses existing `markdown-it` extensions.

To get this counting mechanisms you can

- either make a `markdown-it` extension, like the `markdown-it-footnote` mentioned above,
  (_This changes how Markdown is rendered **internally**_)
- or write an **external** script, which doesn't care about Markdown syntax but only transforms some user-defined syntax (e.g. `[equation-id]`) to KaTeX `\tag` command. (just like a preprocessor)

**Comment by berkdemir** [comment]
\tag {1} works, however, would it be easy to change it to \tag{auto} for automatic numbering?

**Comment by yzhang-gh** [comment]
That is what I mean an **external** script (preprocessor). Someone can make a VS Code extension which reads the current Markdown file and replaces all the `\tag{*}` with increasing indices. Every time you change something, you can run a command like "update equation numbering" to get the correct `\tag`.
